algorithms/kernel/dtrees/forest/regression/oneapi/cl_kernels/df_batch_regression_kernels.cl

```diff
@@ -0,0 +1,994 @@
+        buf[local_id * nProp + 1] = (algorithmFPType)0;
+        buf[local_id * nProp + 2] = (algorithmFPType)0;
+
+        algorithmFPType mrgN        = (algorithmFPType)0;
+        algorithmFPType mrgMean     = (algorithmFPType)0;
+        algorithmFPType mrgSum2Cent = (algorithmFPType)0;
+
+        const __global algorithmFPType * nodePartialHistograms =
+            partialHistograms + nodeIdx * nPartialHistograms * nSelectedFeatures * nMaxBinsAmongFtrs * nProp;
+        __global algorithmFPType * nodeHistogram = histograms + nodeIdx * nSelectedFeatures * nMaxBinsAmongFtrs * nProp;
+
+        for (int i = local_id; i < nPartialHistograms; i += local_size)
+        {
+            int offset        = i * nSelectedFeatures * nMaxBinsAmongFtrs * nProp + binId * nProp;
+            algorithmFPType n = nodePartialHistograms[offset + 0];
+
+            if ((algorithmFPType)0 == n) continue;
+
+            algorithmFPType mean     = nodePartialHistograms[offset + 1];
+            algorithmFPType sum2Cent = nodePartialHistograms[offset + 2];
+
+            // will replace with mergeStat func call
+            algorithmFPType sumN1N2    = mrgN + n;
+            algorithmFPType mulN1N2    = mrgN * n;
+            algorithmFPType deltaScale = mulN1N2 / sumN1N2;
+            algorithmFPType meanScale  = (algorithmFPType)1 / sumN1N2;
+            algorithmFPType delta      = mean - mrgMean;
+
+            mrgSum2Cent = mrgSum2Cent + sum2Cent + delta * delta * deltaScale;
+            mrgMean     = (mrgMean * mrgN + mean * n) * meanScale;
+            mrgN        = sumN1N2;
+
+            buf[local_id * nProp + 0] += n;
+            buf[local_id * nProp + 1] = mrgMean;
+            buf[local_id * nProp + 2] = mrgSum2Cent;
+        }
+
+        for (int offset = local_size / 2; offset > 0; offset >>= 1)
+        {
+            barrier(CLK_LOCAL_MEM_FENCE);
+            if (local_id < offset)
+            {
+                algorithmFPType n = buf[(local_id + offset) * nProp + 0];
+                if ((algorithmFPType)0 == n) continue;
+                algorithmFPType mean     = buf[(local_id + offset) * nProp + 1];
+                algorithmFPType sum2Cent = buf[(local_id + offset) * nProp + 2];
+
+                // will replace with mergeStat func call
+                algorithmFPType sumN1N2    = mrgN + n;
+                algorithmFPType mulN1N2    = mrgN * n;
+                algorithmFPType deltaScale = mulN1N2 / sumN1N2;
+                algorithmFPType meanScale  = (algorithmFPType)1 / sumN1N2;
+                algorithmFPType delta      = mean - mrgMean;
+
+                mrgSum2Cent = mrgSum2Cent + sum2Cent + delta * delta * deltaScale;
+                mrgMean     = (mrgMean * mrgN + mean * n) * meanScale;
+                mrgN        = sumN1N2;
+
+                // item 0 collects all results in private vars
+                // but all others need to store it
+                if (local_id > 0)
```

Comment:
You can actually remove condition here and do the same for local_id=0 as well.
Branching itself is worse than few additional operations made by 0-th thread. Moreover, 0-th thread will still wait for other threads to make this operations.

Reply:
Also, you can avoid store-to-load and put results directly into `buf`.

Reply:
removed local_id check, for second comment - i especially make final write with using private variables to escape one more read from local mem